substrate/client/transaction-pool/src/graph/pool.rs

```diff
@@ -316,17 +318,31 @@ impl<B: ChainApi, L: EventHandler<B>> Pool<B, L> {
 		let in_pool_hashes =
 			extrinsics.iter().map(|extrinsic| self.hash_of(extrinsic)).collect::<Vec<_>>();
 		let in_pool_tags = self.validated_pool.extrinsics_tags(&in_pool_hashes);
+		// Fill unknown tags based on the known tags given in `known_provides_tags`.
+		let mut unknown_txs_count = 0usize;
+		let mut reused_txs_count = 0usize;
+		let tags = in_pool_hashes.iter().zip(in_pool_tags).map(|(tx_hash, tags)| {
+			tags.or_else(|| {
+				unknown_txs_count += 1;
+				known_provides_tags.as_ref().and_then(|inner| {
+					inner.get(&tx_hash).map(|found_tags| {
+						reused_txs_count += 1;
+						found_tags.clone()
+					})
+				})
+			})
+		});
```

Comment:
nit: `known_provides_tags.as_ref().and_then` could be moved outside the loop. So when `None` is passed we do not do useless iterations  over the `in_pool_hashes`.

Reply:
This is a `Map` iterator which we zip with the raw extrinsics, and it is not collected. Down the line we iterate over it in lockstep with each raw extrinsic, in the for loop where we also validate the extrinsic if we don't have its provides tags. It doesn't produce useless iterations, it just executes some logic with each extrinsic, which was already the case by now.

Reply:
My bad, unfortunate wording. You check `inner.is_none`  (via `map`) for every empty tag. This could be re-orged, so we check `is_none` only once. Basically `.map(|(tx_hash, tags)| ...` is not needed if `known_provides_tags` is None.